src/server/editorServices.ts

```diff
@@ -2136,12 +2190,14 @@ namespace ts.server {
             if (lastFileExceededProgramSize) {
                 project.disableLanguageService(lastFileExceededProgramSize);
                 project.stopWatchingWildCards();
+                this.removeProjectFromSharedExtendedConfigFileMap(project);
             }
             else {
                 project.setCompilerOptions(compilerOptions);
                 project.setWatchOptions(parsedCommandLine.watchOptions);
                 project.enableLanguageService();
                 project.watchWildcards(new Map(getEntries(parsedCommandLine.wildcardDirectories!))); // TODO: GH#18217
+                this.updateSharedExtendedConfigFileMap(project);

```

Comment:
You would want to send oldOptions.configFile.extendedSourceFiles list to make sure the remove happens correctly.
You also need test case for scenario where you are removing extended source file and that watcher gets removed.

Reply:
Where would I get `oldOptions`? Collect it from `project.getCompilerOptions()` before it's replaced by `project.setCompilerOptions(compilerOptions)`?

Reply:
Yes. Getting `project.getCompilerOptions()` before https://github.com/microsoft/TypeScript/pull/41493/files#diff-501dedafe824c2480f7f02a2cb5295a90d81a09b45aafaa135b739b51b108cb6R2196 where it sets new options

Reply:
Sorry, removed question. Sounds good.

Reply:
Was going to use oldOptions, but instead I changed the shared update functions (for both tsbuild/tsserver) to remove any reference of a project from unrelated tsconfigs. So now after adding new watchers and updating existing ones, it loops over all watchers and if it's not referenced by the project, that project is removed from the watcher.